Start documenting the configuration file format (#1946)
getsops/sops: README.rst

                azure_keyvault:
                  - https://vault.url/keys/key-name/key-version

Key group object
++++++++++++++++

A key group contains multiple identities (keys), similar to a creation rule object.
Having more than one key group allows to use `Shamir's secret sharing <https://en.wikipedia.org/wiki/Shamir%27s_secret_sharing>`__
to split the file's encryption key up into multiple parts,
requiring more than one identity to access the file.

A key group supports the following keys:

* ``kms`` (list of objects): list of AWS master keys.
  Every object must have the following keys:

  * ``arn`` (string): the ARN of the master key.

  * ``role`` (string, can be omitted): the key's role.

  * ``context`` (object mapping strings to strings): the key's encryption context.

  * ``aws_profile`` (string): the AWS profile.

  Example:

  .. code:: yaml

    - arn: arn:aws:kms:us-west-2:927034868273:key/fe86dd69-4132-404c-ab86-4269956b4500
      role: arn:aws:iam::927034868273:role/sops-dev-xyz
      context:
        Environment: production
        Role: web-server
      aws_profile: foo

> * ``gcp_kms`` (list of objects): list of GCP KMS ResourceIDs.

Author: I got the impression from https://github.com/getsops/sops/blob/44e07cd9cc190a4d67f07cecdeaea28e8f97f342/README.rst#24encrypting-using-gcp-kms that `ResourceID` is a fixed term. On the other hand, Google talks about "resource IDs" for example in https://cloud.google.com/kms/docs/getting-resource-ids. So I guess it should be:
```suggestion
* ``gcp_kms`` (list of objects): list of GCP KMS resource IDs.
```
(I didn't add the apostrophe since we don't have a style guide that mandates it (to my knowledge), and https://english.stackexchange.com/questions/212040/is-the-plural-form-of-id-spelled-ids-or-id :) )

  * ``key`` (string): the key name.

  * ``version`` (string, can be empty): the version of the key to use.
    If empty, the latest key will be used on encryption.

  Example:

  .. code:: yaml

    - vaultUrl: https://vault.url
      key: key-name
      version: key-version
    - vaultUrl: https://vault.url
      key: key-name
      version: ""

* ``hc_vault`` (list of strings): list of HashiCorp Vault transit URIs.

* ``age`` (list of strings): list of Age public keys.

* ``pgp`` (list of strings): list of PGP/GPG key fingerprints.

* ``merge``: a list of key group objects.
  These will be merged (by concatenating the keys of the same type) into this key group.
  This key is only there to allow concatenation of key groups using YAML anchors, aliases, and overrides.

Settings
********

The following keys configure encryption settings:

* ``shamir_threshold`` (integer, default ``0``): Must be ``0`` (disabled) or an integer greater or equal to 2.
  Determines the number of key groups from whose one key must be present each to decrypt the file's key.

* ``mac_only_encrypted`` (boolean, default ``false``): If set to ``true``, only encrypted strings will count towards the file's MAC.

>   If set to ``false``, also unencrypted values will be part of the MAC computation.

Reviewer: ```suggestion
  If set to ``false``, unencrypted values will also be part of the MAC computation.
```

> The following keys configure which values in a file are encrypted.

Reviewer: ```suggestion
The following keys configure the specific values in a file that should be encrypted.
```

> Note that at most one of these keys can be used.

Reviewer: ```suggestion
Note that at most, one of these keys can be used.
```

  All other values are **encrypted**.
  Comments are always encrypted.

* ``encrypted_suffix`` (string): A value is encrypted if its key **does** end with this suffix.
  All other values are **not encrypted**.
  Comments are always encrypted.

* ``unencrypted_regex`` (string): A value is encrypted if its key **does not match** this regular expression.
  All other values are **encrypted**.
  Comments are always encrypted.

* ``encrypted_regex`` (string): A value is encrypted if its key **matches** this regular expression.
  All other values are **not encrypted**.
  Comments are always encrypted.

* ``unencrypted_comment_regex`` (string): A value or comment is not encrypted if it has a preceding comment (or a trailing comment on the same line) matching this regular expression.
  All other values and comments are encrypted.

* ``encrypted_comment_regex`` (string): A value or comment is encrypted if it has a preceding comment (or a trailing comment on the same line) matching this regular expression.
  All other comments and values are not encrypted.
  The matching comment itself is not encrypted.

Destination rule object
~~~~~~~~~~~~~~~~~~~~~~~

Not yet documented.

> .. TODO!

Reviewer: Lets remove the TODO line as we already have a callout for the section of the document that is lacking content

Reviewer: It can stay. RST is indeed "fun"
